aten/src/ATen/native/NNPACK.cpp

```diff
@@ -56,53 +56,39 @@ bool _nnpack_available() {
 
 #include "nnpack.h"
 
-#include <stdlib.h>
-
-#include <ATen/Parallel.h>
-#include <thread>
+#include "caffe2/utils/threadpool/ThreadPoolMobile.h"
 
 namespace at {
 namespace native {
 
-// Stolen from Caffe2
-static pthreadpool_t nnpack_threadpool_ = nullptr;
-static bool called_nnpack_threadpool_ = false;
+static bool init_nnpack() {
+  static std::once_flag once;
+  static bool nnpack_successfully_initialized = false;
+
+  std::call_once(once, []() {
+    const nnp_status nnpack_status = nnp_initialize();
+    nnpack_successfully_initialized = (nnp_status_success == nnpack_status);
 
-pthreadpool_t nnpack_threadpool() {
-  if (! called_nnpack_threadpool_) {
-    called_nnpack_threadpool_ = true;
-    enum nnp_status nnpack_status = nnp_initialize();
     if (nnpack_status != nnp_status_success) {
       if (nnpack_status == nnp_status_out_of_memory) {
-        throw std::runtime_error("could not initialize NNPack (out of memory)");
+        LOG(WARNING) << "Could not initialize NNPACK! Reason: Out of memory.";
       } else if (nnpack_status == nnp_status_unsupported_hardware) {
-        throw std::runtime_error("could not initialize NNPack (unsupported hardware)");
+        LOG(WARNING) << "Could not initialize NNPACK! Reason: Unsupported hardware.";
       } else {
-        throw std::runtime_error("could not initialize NNPack (unknown error)");
+        LOG(WARNING) << "could not initialize NNPACK! Reason: Unknown error!";
       }
     }
-    unsigned int threads;
-#ifdef INTRA_OP_PARALLEL
-    threads = at::get_num_threads();
-#else
-    threads = std::thread::hardware_concurrency();
-#endif
-    nnpack_threadpool_ = pthreadpool_create(threads);
-    if (nnpack_threadpool_ == nullptr) {
-      throw std::runtime_error("could not initialize NNPack's pthreadpool");
-    }
-  }
-  return nnpack_threadpool_;
+  });
+
+  return nnpack_successfully_initialized;
+}
+
+static pthreadpool_t nnpack_threadpool() {
+  return caffe2::mobile_pthreadpool();
```

Comment:
mobile_pthreadpool() returns null for non-mobile build.
I don't know if NNPACK is used for non-mobile build at all - maybe @ezyang / @ilia-cher knows?
if we want to keep 100% the same behavior we should fallback to old thread creation logic for non-mobile build...

Reply:
Right, valid point.  I'll wait for Illia's input.

Reply:
This test suggest nnpack is used in non-mobile:

```
    @unittest.skipIf(not torch._nnpack_available(), "NNPACK unavailable")
    def test_nnpack_conv(self):
        for kern, inp_size in [(3, 6), (3, 7), (4, 9)]:
            for batch, padding, chan_in, chan_out in \
                    product([1, 2], [0, 1, 2], [2], [3]):
```

Reply:
let's fallback to a non-mobile thread pool